bam/writer.go

```diff
@@ -107,83 +109,72 @@ func (bw *Writer) Write(r *sam.Record) error {
 	bin.writeInt32(int32(r.TempLen))
 
 	// Write variable length data.
-	wb.Write(append([]byte(r.Name), 0))
+	bin.writeString(r.Name)
+	bin.writeByte(0)
 	writeCigarOps(&bin, r.Cigar)
-	wb.Write(doublets(r.Seq.Seq).Bytes())
+	bin.write(doublets(r.Seq.Seq).Bytes())
 	if r.Qual != nil {
-		wb.Write(r.Qual)
+		bin.write(r.Qual)
 	} else {
 		for i := 0; i < r.Seq.Length; i++ {
-			wb.WriteByte(0xff)
+			bin.writeByte(0xff)
 		}
 	}
-	wb.Write(tags)
-	if wb.err != nil {
-		return wb.err
-	}
-
+	bin.write(tags)
 	_, err := bw.bg.Write(bw.buf.Bytes())
 	return err
 }
 
 func writeCigarOps(bin *binaryWriter, co []sam.CigarOp) {
 	for _, o := range co {
 		bin.writeUint32(uint32(o))
-		if bin.w.err != nil {
-			return
-		}
 	}
-	return
 }
 
 // Close closes the writer.
 func (bw *Writer) Close() error {
 	return bw.bg.Close()
 }
 
-type errWriter struct {
-	w   *bytes.Buffer
-	err error
+func (w *binaryWriter) write(p []byte) {
+	if _, err := w.w.Write(p); err != nil {
```

Comment:
~These are documented to not be necessary as you mention, so don't bother with this.~